### PR TITLE
Addon Test: Always run Vitest in watch mode internally

### PR DESCRIPTION
**Pull request by JReinhold**
## What I did

Changed the Vitest logic to _always_ run in watch mode, and then internally disable triggering tests on file changes when SB watch mode is disabled. This ensures that coverage is always re-intrumented on file changes, and Vitest module caches, even when SB watch mode is disabled.

Basically, when we want the flow where we're re-using the Vitest instance, we want to always enable watch mode too, as that is how Vitest expects to behave. Disabled watch-mode are only supported for single-use Vitest instances.

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [x] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.7 MB | 77.7 MB | 1.48 kB | -0.3 | 0% |
| initSize |  133 MB | 133 MB | 1.5 kB | **2.97** | 0% |
| diffSize |  55.1 MB | 55.1 MB | 24 B | **2.95** | 0% |
| buildSize |  6.87 MB | 6.87 MB | 0 B | 1.16 | 0% |
| buildSbAddonsSize |  1.51 MB | 1.51 MB | 0 B | **3** | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.86 MB | 1.86 MB | 0 B | **2.95** | 0% |
| buildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.57 MB | 3.57 MB | 0 B | **3** | 0% |
| buildPreviewSize |  3.3 MB | 3.3 MB | 0 B | 1.13 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  7.4s | 7.3s | -122ms | -0.54 | -1.7% |
| generateTime |  21.5s | 21.2s | -317ms | 0.36 | -1.5% |
| initTime |  13.5s | 15.5s | 1.9s | **1.62** | 🔺12.6% |
| buildTime |  8.3s | 12.2s | 3.8s | **5.99** | 🔺31.4% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  5.7s | 5.4s | -268ms | 0.38 | -4.9% |
| devManagerResponsive |  4.2s | 4s | -234ms | 0.61 | -5.8% |
| devManagerHeaderVisible |  576ms | 692ms | 116ms | **1.76** | 🔺16.8% |
| devManagerIndexVisible |  677ms | 707ms | 30ms | 0.54 | 4.2% |
| devStoryVisibleUncached |  1.9s | 1.9s | 12ms | 0.63 | 0.6% |
| devStoryVisible |  673ms | 731ms | 58ms | 1.1 | 7.9% |
| devAutodocsVisible |  686ms | 462ms | -224ms | -0.9 | -48.5% |
| devMDXVisible |  562ms | 677ms | 115ms | **2.28** | 🔺17% |
| buildManagerHeaderVisible |  748ms | 590ms | -158ms | 0.02 | -26.8% |
| buildManagerIndexVisible |  757ms | 735ms | -22ms | 0.66 | -3% |
| buildStoryVisible |  596ms | 518ms | -78ms | -0.16 | -15.1% |
| buildAutodocsVisible |  426ms | 472ms | 46ms | 0.32 | 9.7% |
| buildMDXVisible |  510ms | 431ms | -79ms | -0.22 | -18.3% |

## Greptile Summary

Modified Vitest test runner to always operate in watch mode internally while using Storybook's watch mode flag to control test execution, improving cache invalidation and coverage instrumentation.

- Changed `VitestManager` to always use `watch: true` in Vitest configuration
- Added guard in `runAffectedTestsAfterChange` to only trigger tests when Storybook watch mode is enabled
- Set `cleanOnRerun: false` for coverage options to preserve instrumentation state
- Updated `TestManager` to handle coverage mode changes without affecting watch mode state
- Modified test suite to verify new watch mode and coverage behavior

### CODE REVIEW

File: code/addons/test/src/node/vitest-manager.ts [hunk 1]
```diff
@@ -276,6 +273,9 @@ export class VitestManager {
     this.updateLastChanged(id);
     this.storyCountForCurrentRun = 0;
 
+    if (!this.testManager.watchMode) {
+      return;
+    }
```
Comment (JReinhold) [inline code/addons/test/src/node/vitest-manager.ts:278]: The key was to return _here_, _after_ we've invalidated caches and whatnot above.

Initially I returned immediately in this function, but that created wrong coverage results too, so we have to return just before running tests.

**Review by valentinpalkovic — APPROVED** [review]

File: code/addons/test/src/node/test-manager.ts [hunk 2]
```diff
@@ -52,24 +52,28 @@ export class TestManager {
...
+    }
+    this.watchMode = payload.watchMode;
 
-      if (this.watchMode !== payload.watchMode) {
-        this.watchMode = payload.watchMode;
-        await this.restartVitest({ watchMode: this.watchMode, coverage: false });
+    try {
+      if (payload.watchMode && this.coverage) {
+        // if watch mode is toggled on and coverage is already enabled, restart vitest without coverage to automatically disable it
+        this.restartVitest({ coverage: false });
```
Comment (ghengeveld) [inline code/addons/test/src/node/test-manager.ts:73]: Shouldn't this be awaited?

File: code/addons/test/src/node/test-manager.ts [hunk 3]
```diff
@@ -52,24 +52,28 @@ export class TestManager {
...
+      } else if (!payload.watchMode && this.coverage) {
+        // if watch mode is toggled off and coverage is already enabled, restart vitest with coverage to automatically re-enable it
+        this.restartVitest({ coverage: this.coverage });
```
Comment (ghengeveld) [inline code/addons/test/src/node/test-manager.ts:76]: Shouldn't this be awaited as well?

File: code/addons/test/src/node/test-manager.ts [hunk 4]
```diff
@@ -82,28 +86,22 @@ export class TestManager {
...
-           among other things it causes the coverage report to be incorrect and stale.
-           
-           If we're only running a subset of stories, we have to temporarily disable coverage,
-           as a coverage report for a subset of stories is not useful.
-         */
+      /*
+        If we're only running a subset of stories, we have to temporarily disable coverage,
+        as a coverage report for a subset of stories is not useful.
+      */
+      const temporarilyDisableCoverage = this.coverage && (payload.storyIds ?? []).length === 0;
```
[the remark this answers is not in the document]
Comment (ghengeveld) [reply code/addons/test/src/node/test-manager.ts:96]: I second that. Seems wrong.

File: code/addons/test/src/node/vitest-manager.ts (hunk 1 above)
Comment (ghengeveld) [reply code/addons/test/src/node/vitest-manager.ts:278]: Maybe add a comment about the relevancy of returning here.

**Review by ghengeveld — COMMENTED** [review]
Needs a bit of work, I think.

**Review by ghengeveld — APPROVED** [review]

**Pull request merged**